File: Android/app/src/main/java/com/example/android/canvas/Selection.kt
package com.example.android.canvas

import android.content.Context
import android.view.View
import com.example.android.client.ClientInfo
import org.apache.batik.anim.dom.SVGOMGElement
import org.apache.batik.dom.AbstractDocument
import org.w3c.dom.Element
import org.w3c.dom.svg.SVGElement

class Selection(private var drawingId: Int?, private var owner: AbstractDocument)
    : Tool, SVGOMGElement()
{
    override var currentX = 0f
    override var currentY = 0f
    override var str = ""
    override var selected = false
    override var startTransformPoint = Point(0f,0f)
    override var totalTranslation = Point(0f,0f)
    override var totalScaling = Point(0f,0f)
    override var scalingPositions = HashMap<Point, Point>()
    override var contentID: Int? = null
    override var userId: String? = ClientInfo.userId
    private var selectedTool: Tool? = null


    override fun touchStart(eventX: Float, eventY: Float, svgRoot: Element) {
        // Take the most recent element on the stack
        var i = svgRoot.childNodes.length - 1
        if(svgRoot.childNodes.length > 0){
            while(i >= 0){
                val tool = svgRoot.childNodes.item(i) as Tool
                if(tool.inTranslationZone(eventX, eventY)){
                    if(tool.userId == null ||
                        !(tool.selected && tool.userId != ClientInfo.userId)){
                        if (tool is FreeHand){
                            selectedTool = FreeHand(drawingId, pencilString, owner)
                        }
                        else if(tool is Ellipse){
                            selectedTool = Ellipse(drawingId, ellipseString, owner)
                        }
                        else{
                            selectedTool = Rectangle(drawingId, rectString, owner)
                        }
                        selectedTool!!.parse(tool.getOriginalString())
                        startTransformPoint.x = tool.startTransformPoint.x
                        startTransformPoint.y = tool.startTransformPoint.y
                        selectedTool!!.contentID = tool.contentID
                        selectedTool!!.select()
                    }

                    break
                }
                i--
            }
        }
    }

    override fun touchMove(context: Context, eventX: Float, eventY: Float) {
        //Not needed given the exception that will be made in the CanvasView
    }

    fun getSelectedTool(): Tool?{
        return selectedTool
    }
    override fun touchUp() {
        //setCriticalValues()
        if(selectedTool!=null){
            selectedTool!!.touchUp()
        }
        //calculateScalingPositions()
    }

    override fun getString(): String { return str}

    override fun getOriginalString(): String { return str}

    override fun inTranslationZone(eventX: Float, eventY: Float): Boolean {
        if(selectedTool != null){
            return selectedTool!!.inTranslationZone(eventX, eventY)
        }
        return false
    }

    override fun translate(view: View, translationPoint: Point) {
        if(selectedTool != null){
            selectedTool!!.translate(view, translationPoint)
        }
    }

    override fun scale(view: View, scalePoint: Point, direction: Point) {
        if(selectedTool != null){
            selectedTool!!.scale(view, scalePoint, direction)
        }
    }

    override fun getSelectionString() {}

    override fun calculateScalingPositions() {
        if(selectedTool != null){
            selectedTool!!.calculateScalingPositions()
        }
    }

    override fun getScalingPoint(point: Point): MutableMap.MutableEntry<Point, Point>? {
        if(selectedTool != null){
            return selectedTool!!.getScalingPoint(point)
        }
        return null
    }

    override fun getScalingPositionsString() {}

    override fun parse(parceableString: String?) {}

    override fun unselect() {
        if(selectedTool != null){
            selectedTool!!.unselect()
        }
    }

    override fun delete() {
        if(selectedTool != null){
            selectedTool!!.delete()
            selectedTool = null
        }
    }

    override fun updateThickness() {
        if(selectedTool != null){
            selectedTool!!.updateThickness()
        }
    }

    override fun updatePrimaryColor() {
        if(selectedTool != null){
            selectedTool!!.updatePrimaryColor()
        }
    }

    override fun updateSecondaryColor() {
        if(selectedTool != null){
            selectedTool!!.updateSecondaryColor()
        }
    }
    override fun select(){}
    override fun setCriticalValues() {
        if(selectedTool!= null){
            selectedTool!!.setCriticalValues()
        }
    }
}

File: angular-client/src/app/services/drawing-tools/selection.ts
import { ElementRef, Renderer2 } from '@angular/core';
import { ColorPickingService } from '../color-picker/color-picking.service';
import { DrawingTool } from './drawing-tool';
import { Point } from './point';
import { InteractionService } from '../interaction/interaction.service';
import { SocketService } from '../socket/socket.service';
import { Pencil } from './pencil';
import { Rectangle } from './rectangle';
import { Ellipse } from './ellipse';
import { DrawingService } from '../drawing/drawing.service';

//const INIT_VALUE = -1;

export class Selection implements DrawingTool {

  currentX: number = 0.0;
  currentY: number=0.0;
  str: string= "";
  selected: boolean = false;;
  startTransformPoint: Point = new Point(0,0);
  totalTranslation: Point = new Point(0,0);
  totalScaling: Point = new Point(0,0);
  scalingPositions: Map<Point, Point> = new Map<Point, Point>();
  contentId!: number;
  userId: string;
  element!: SVGElement;
  drawingId!: number;
  toolFound: boolean = false;
  //mouseIsDown: boolean = false;
  selectedTool: DrawingTool | undefined;
  constructor(
    public toolsArray: Map<number,DrawingTool>,
    private socketSerice: SocketService,
    private colorPick: ColorPickingService,
    private renderer: Renderer2,
    private canvas: ElementRef,
    private interactionService: InteractionService,
    drawingId: number,
    //private drawingId: number,
    userId: string,
    private drawingService: DrawingService,
    ) {
        this.drawingId = drawingId;
        this.userId = userId;
    }
    onMouseDown(event: MouseEvent): void {
        this.toolsArray.forEach((tool: DrawingTool)=>{
            if(!this.toolFound && tool.inTranslationZone(event)){
                if(tool.userId === null || !(tool.selected && tool.userId !== this.userId)){
                    if(tool instanceof Pencil){
                        this.selectedTool = new Pencil(this.interactionService, this.colorPick, this.socketSerice, this.userId, this.renderer, this.canvas, this.drawingService);
                    }
                    else if(tool instanceof Rectangle){
                        this.selectedTool = new Rectangle(this.interactionService, this.colorPick, this.socketSerice, this.userId, this.renderer, this.canvas, this.drawingService);
                    }
                    else{
                        this.selectedTool = new Ellipse(this.interactionService, this.colorPick, this.socketSerice, this.userId, this.renderer, this.canvas, this.drawingService)
                    }
                    this.selectedTool.drawingId = this.drawingId;
                    this.selectedTool.parse(tool.getOriginalString())
                    this.startTransformPoint.x = tool.startTransformPoint.x;
                    this.startTransformPoint.y  = tool.startTransformPoint.y;
                    this.selectedTool.contentId = tool.contentId;
                    this.selectedTool.select();
                    this.toolFound = true;
                }
                
            }
        })
    }
    onMouseUp(e: MouseEvent): void {
        if(this.selectedTool!== undefined){
            this.selectedTool.onMouseUp(e);
        }
    }
    getSelectedTool():DrawingTool | undefined{
        return this.selectedTool;
    }
    onMouseMove(event: MouseEvent): void {
    }
    getString(): string {

        return this.str;
    }
    getOriginalString(): string {

        return this.str;
    }
    inTranslationZone(event: MouseEvent): boolean {
        if(this.selectedTool!== undefined){
            return this.selectedTool.inTranslationZone(event);
        }
        return false;

    }
    translate(translationPoint: Point): void {
        if(this.selectedTool!== undefined){
            this.selectedTool.translate(translationPoint);
        }

    }
    scale(scalePoint: Point, direction: Point): void {
        if(this.selectedTool!== undefined){
            this.selectedTool.scale(scalePoint, direction);
        }

    }
    getSelectionString(): void {

    }
    calculateScalingPositions(): void {
        if(this.selectedTool!== undefined){
            this.selectedTool.calculateScalingPositions();
        }
 
    }
    getScalingPoint(point: Point): [Point, Point] | null {
        if(this.selectedTool!== undefined){
            return this.selectedTool.getScalingPoint(point);
        }
        return null;

    }
    getScalingPositionsString(): void {

    }
    parse(parceableString: string): void {
  
    }
    unselect(): void {
 
        if(this.selectedTool!== undefined){
            this.selectedTool.unselect();
        }
    }
    delete(): void {
        if(this.selectedTool!== undefined){
            this.selectedTool.delete();
            this.selectedTool = undefined;
        }

    }
    updateThickness(): void {
 
        if(this.selectedTool !== undefined){
            this.selectedTool.updateThickness();
        }
    }
    updatePrimaryColor(): void {

        if(this.selectedTool !== undefined){
            this.selectedTool.updatePrimaryColor();
        }
    }
    updateSecondaryColor(): void {

        if(this.selectedTool !== undefined){
            this.selectedTool.updateSecondaryColor();
        }
    }
    select(): void {

    }
    setCriticalValues(): void {
        if(this.selectedTool!== undefined){
            this.selectedTool.setCriticalValues();
        }

    }
      
}
